Add @UpdateForV10 to AbstractObjectParser methods

joshua-adams-1 opened the pull request:
Adds an `@UpdateForV10` annotation to the following methods from the `AbstractObjectParser` class:
- `declareInt`
- `declareIntOrNull`
- `declareLong`
- `declareLongOrNull`
- `declareIntArray`
- `declareLongArray`

 and assigns it to the Core Infra team.  The required change is explained in more detail in #130797. These current parsing methods do not throw exceptions when non-integer, or non-long, values, such as floats, doubles and stringified floats and doubles, are passed. The current behaviour is to truncate the floats / doubles into ints / longs. This is problematic for two reasons:

1. Customers may depend on this implicit behaviour which is not officially supported
2. Parameters passed via query parameters have stricter validation on floats, and there is a discrepancy between which values are supported via each method

The proposed solution will be to throw `IllegalArgumentException`'s whenever non-integer and non-long values are passed.

joshua-adams-1 committed:
Add @UpdateForV10 to AbstractObjectParser.declareInt

Adds an `@UpdateForV10` annotation to the `AbstractObjectParser
.declareInt` method and assigns it to the Core Infra team.
 This method needs to throw an `IllegalArgumentException` for
 non-integer values being passed, such as floats, doubles, and
 stringified floats and doubles.

joshua-adams-1 committed:
Merge branch 'main' into declare-int-update-v10

DaveCTurner reviewed (commented):
I guess the same also applies to `declareLong` and the `...OrNull` variants thereof?

joshua-adams-1 committed:
Add @UpdateForV10 to AbstractObjectParser methods

Adds an `@UpdateForV10` annotation to the following methods in
`AbstractObjectParser`: `declareInt`, `declareIntOrNull`,
`declareLong`,
 `declareLongOrNull`, `declareIntArray` and `declareLongArray`.

 The required change is explained in https://github
 .com/elastic/elasticsearch/issues/130797. The current parsing methods
 do not throw exceptions when floats or doubles are passed in place of
 longs or ints. The current behaviour is to truncate the float /
 double but this can lead to incorrect behaviour, and is also a
 discrepancy versus the stricter validation found when floats are passed
  as query parameters.

joshua-adams-1 committed:
Merge branch 'main' into declare-int-update-v10

joshua-adams-1 commented:
@DaveCTurner Good catch! Have expanded the PR accordingly

joshua-adams-1 committed:
Merge branch 'main' into declare-int-update-v10

DaveCTurner reviewed (approved):
LGTM
